Batteries/Data/List/Count.lean

```diff
@@ -25,8 +26,42 @@ namespace List
 
 section count
 
-variable [DecidableEq α]
+theorem count_singleton' [DecidableEq α] (a b : α) : count a [b] = if b = a then 1 else 0 :=
+  count_singleton.trans (by grind)
 
-theorem count_singleton' (a b : α) : count a [b] = if b = a then 1 else 0 := by simp [count_cons]
+variable [BEq α]
 
-theorem count_concat (a : α) (l : List α) : count a (concat l a) = succ (count a l) := by simp
+theorem count_concat [LawfulBEq α] (a : α) (l : List α) :
+    count a (concat l a) = succ (count a l) := by simp
+
+/-- For a fixed list, given the index of an element,
+  find the corresponding element and multiplicity. -/
+def idxToCountElem [EquivBEq α] (xs : List α) (i : Fin xs.length) :
+    (x : α) × Fin (xs.count x) := ⟨xs[i], (xs.idxsOf xs[i]).idxOf ↑i, by grind⟩
```

Comment:
Computing `idxsOf` only to truncate is wasteful here. You should count the number that occur before position `i` instead.